docs/API_DESIGN.md

```diff
@@ -0,0 +1,658 @@
+  PUBLIC DATA capabilityDelegation: []string?
+
+  /// Services are used in DID documents to express ways of communicating with
+  /// the DID subject or associated entities.
+  /// A service can be any type of service the DID subject wants to advertise.
+  ///
+  /// [Specification Reference](https://www.w3.org/TR/did-core/#services)
+  PUBLIC DATA service: []Service?
+```
+
+### `VerificationMethod`
+
+```pseudocode!
+/// Representation of a DID Document's [Verification Method](https://www.w3.org/TR/did-core/#verification-methods).
+CLASS VerificationMethod
+  /// 🚧
+  PUBLIC DATA id: string
+
+  /// 🚧
+  PUBLIC DATA type: string
+
+  /// 🚧
+  PUBLIC DATA controller: string
+
+  /// 🚧
+  PUBLIC DATA publicKeyJwk: Jwk
+```
+
+### `Service`
+
+```pseudocode!
+/// Representation of a DID Document's [Service](https://www.w3.org/TR/did-core/#service).
+CLASS Service
+  /// 🚧
+  PUBLIC DATA id: string
+
+  /// 🚧
+  PUBLIC DATA type: string
+
+  /// 🚧
+  PUBLIC DATA serviceEndpoint: []string
+```
+
+## `ResolutionResult`
+
+```pseudocode!
+/// Representation of the result of a DID (Decentralized Identifier) resolution.
+CLASS ResolutionResult
+  /// The resolved DID document, if available.
+  PUBLIC DATA document: Document
+
+  /// The metadata associated with the DID document.
+  PUBLIC DATA document_metadata: DocumentMetadata
+
+  /// The metadata associated with the DID resolution process.
+  PUBLIC DATA resolution_metadata: ResolutionMetadata
+
+  /// Resolve via a DID URI.
+  CONSTRUCTOR resolve(uri: string)
+```
+
+**Examples**
+
+```pseudocode!
+🚧
+```
+
+### `ResolutionMetadataError`
+
+```pseudocode!
+/// The error code from the resolution process.
+ENUM ResolutionMetadataError
+  invalidDid
+  notFound
+  representationNotSupported
+  methodNotSupported
+  invalidDidDocument
+  invalidDidDocumentLength
+  internalError
+```
+
+### `ResolutionMetadata`
+
+```pseudocode!
+/// Metadata about the given resolution.
+CLASS ResolutionMetadata
+  /// The error code from the resolution process.
+  PUBLIC DATA error: ResolutionMetadataError?
+```
+
+### `DocumentMetadata`
+
+```pseudocode!
+/// Metadata about the DID Document.
+CLASS DocumentMetadata
+  /// 🚧
+  PUBLIC DATA created: string?
+
+  /// 🚧
+  PUBLIC DATA updated: string?
+
+  /// 🚧
+  PUBLIC DATA deactivated: bool?
+
+  /// 🚧
+  PUBLIC DATA nextUpdate: string?
+
+  /// 🚧
+  PUBLIC DATA versionId: string?
+
+  /// 🚧
+  PUBLIC DATA nextVersionId: string?
+
+  /// 🚧
+  PUBLIC DATA equivalentId: []string?
+
+  /// 🚧
+  PUBLIC DATA canonicalId: string?
+```
+
+## Methods
+
+### `DidJwk`
+
+```pseudocode!
+CLASS DidJwk
+  PUBLIC DATA did: Did
+  PUBLIC DATA document: Document
+  CONSTRUCTOR(public_key: Jwk)
+  CONSTRUCTOR(uri: string)
+  STATIC METHOD resolve(uri: string): ResolutionResult
+```
+
+#### Example: Create a `did:jwk`
+
+```pseudocode!
+key_manager = new InMemoryKeyManager()
+public_key = key_manager.generate_key_material()
```

Comment:
can you tell me a bit more about what this function does? would an equivalent in the current impl be `KeyManager#GeneratePrivateKey()` which returns the keyid after storing the private key in the keystore (but now you're returning the public key as the keyid equivalent) ?

Reply:
Yeah it's functionally equivalent

This was subsequently removed in favor of:

```pseudocode!
let key_manager = new InMemoryKeyManager();
let public_jwk = key_manager.import_private_jwk(Ed25519Generator::generate())
```

Reply:
And yeah it's potay-to potah-to with regards to a string representing the thumbprint of the public key; both are functionally equivalent. I chose to go with the JWK being the point of reference because it's one less abstraction to accidentally break.